### PR TITLE
Thread commit SHAs into review-comment prompts

### PR DESCRIPTION
**Pull request by subsetpark**
## Summary

- When the supervisor delivers review comments to the coding agent, the prompt now includes the SHA the review was anchored at and the current branch-tip SHA, plus per-bullet `[at=<short-sha>]` and `[outdated]` annotations. The agent can now distinguish stale feedback from live feedback mechanically instead of guessing.
- Motivation: we observed an agent reply with "already fixed in da442c5" / "you appear to be reviewing 9c759a1" — both wrong. The review had been generated at an earlier SHA; the agent had no anchor to reason against.
- Back-compat: when we have no SHA info (GraphQL returned null, no head fetched), the prompt renders identically to before.

## Changes

- `lib/github.ml`: GraphQL query adds `headRefOid` and per-comment `commit { oid }` / `originalCommit { oid }`; parsers surface them.
- `lib/types.{ml,mli}`: add `commit_sha` / `original_commit_sha` to `Comment.t`.
- `lib/pr_state.{ml,mli}`: add `head_oid` to `Pr_state.t`.
- `lib/prompt.{ml,mli}`: new `?current_head_sha` arg on `render_review_prompt`; emits SHA preamble and per-bullet annotations; adds short-SHA helper.
- `bin/main.ml`: threads `fresh_pr_state.head_oid` through the `Review_payload` render site.
- Tests: updated inline fixtures; added expect-test for the SHA-anchored rendering and a test for the outdated marker.

## Test plan

- [x] `dune build` clean
- [x] `dune runtest` — 18 suites pass; new expect-test captures the anchored output; back-compat test (no SHAs) unchanged
- [x] `dune fmt` clean
- [ ] Manual end-to-end check against a real PR with review comments — confirm rendered prompt contains `Review anchored at commit \`…\``, `[at=…]`, and (when GitHub considers a comment detached) `[outdated]`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/flowglad/codesmith/onton/pr/201"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Codesmith can help with this PR — just tag <code>@codesmith</code> or enable autofix.</sup>

- [ ] Autofix CI and bot reviews

## Summary by CodeRabbit

* **New Features**
  * Review comments now show which commit each comment references (anchored-at short SHAs).
  * When a comment references an earlier commit than the current branch head, it is marked as [outdated].
  * When commit identifiers are missing, comment display and behavior remain unchanged (backward-compatible).
  * Prompt header now includes current branch HEAD when available.

---
## Summary by cubic
Adds commit SHA anchoring to review-comment prompts so the agent can tell which commit feedback targets and when it’s stale. Handles mixed-SHA batches, avoids file-level false positives, and stays back-compatible with clean, single-blank-line formatting.

- **New Features**
  - Fetches `headRefOid` and per-comment `commit.oid` / `originalCommit.oid` via GraphQL (thread opener only).
  - Stores SHAs in `Pr_state.head_oid` and `Comment.commit_sha` / `original_commit_sha`.
  - `render_review_prompt` accepts `?current_head_sha`, emits a preamble for one or more commits, uses `original_commit_sha` for `[at=…]`, and marks `[outdated]` only when SHAs differ (file-level comments are not marked outdated).
  - Mixed-SHA batches: SHA-less comments omit `[at=…]` and are excluded from the preamble.

- **Refactors**
  - Ensures exactly one blank line between header, PR number, and preamble by stripping trailing newlines; normalizes multi-SHA list formatting (", ").
  - Tests converted to `expect_test`s and expanded to cover PR number + head SHA spacing; file-level test hardened; `short_sha` marked private and docs clarified.

<sup>Written for commit b537bd960b6a936e482eb22babffcca8cac5d0fd. Summary will update on new commits.</sup>

### CODE REVIEW

File: lib/prompt.ml [hunk 1]
```diff
@@ -503,6 +505,17 @@ let render_review_prompt ~(project_name : string) ?pr_number
...
       in
+      (* Only claim outdated-ness when GitHub has given us a signal. If
+         [commit] and [originalCommit] differ, GitHub re-anchored the comment
+         to a later commit (line moved). If [line] is null but we do have an
+         original SHA, GitHub detached the comment from the file. With no SHA
+         info at all, say nothing — matches pre-SHA behavior. *)
+      let is_outdated (c : Comment.t) =
+        match (c.Comment.commit_sha, c.Comment.original_commit_sha) with
+        | Some cur, Some orig when not (String.equal cur orig) -> true
+        | _, Some _ when Option.is_none c.Comment.line -> true
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml]: Good catch — fixed in the follow-up commit. I dropped the `line = None with original SHA` clause entirely: it was trying to detect detached comments, but it false-positived on file-level comments (which legitimately have `line = None`). Distinguishing the two would require fetching `subjectType` from GraphQL, which is more invasive; the `commit_sha != original_commit_sha` check already catches re-anchored/outdated comments reliably, so losing the detached-comment edge case is the smaller cost. Added a regression test.

File: lib/prompt.ml [hunk 2]
```diff
@@ -877,3 +938,98 @@ let%test "review prompt formats comments" =
...
+        };
+    ]
+  in
+  let result =
+    render_review_prompt ~project_name:"test"
+      ~current_head_sha:"47525fdaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" comments
+  in
+  (* The preamble references "[outdated]" literally; the bullet-line marker is
+     distinguished by the trailing colon before the body. *)
+  not (String.is_substring result ~substring:"[outdated]:")
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml]: Agreed — swapped to asserting directly on the bullet line. Now picks the line containing "File-level feedback." and checks that that single line has no `[outdated]` substring, so the preamble occurrence can't mask a regression and format-string changes won't silently false-pass.

File: lib/github.ml [hunk 3]
```diff
@@ -206,7 +209,15 @@ let parse_comment_node ~thread_id node =
   let body = node |> member "body" |> to_string in
   let path = node |> member "path" |> to_string_option in
   let line = node |> member "line" |> to_int_option in
-  Types.Comment.{ id; thread_id; body; path; line }
+  let oid_of field =
+    match node |> member field with
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml:220]: Added an explanatory comment documenting the silent-fallback behavior. Didn't add logging — `lib/github.ml` has no structured logger today (no `Logs`, no `Log`, no `eprintf` call sites), so introducing one just for this branch is more scope than the ask. The current [None] fallback is safe (caller treats it as "no anchor"); the comment flags it so the next reader knows to add a warning if richer observability lands.

File: lib/prompt.ml [hunk 4]
```diff
@@ -520,27 +533,71 @@ let render_review_prompt ~(project_name : string) ?pr_number
...
         |> String.concat ~sep:"\n\n"
       in
       let pr_ctx =
         match pr_number with
         | Some n -> Printf.sprintf "\n\nPR: #%d\n" (Pr_number.to_int n)
         | None -> ""
       in
+      let reviewed_at_shas =
+        List.filter_map comments ~f:(fun (c : Comment.t) ->
+            c.Comment.original_commit_sha)
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:569]: Added a comment at the filter site documenting the intentional behavior: mixed batches list only SHA-bearing comments in the preamble, and SHA-less entries simply omit the `[at=…]` annotation. The alternative (fabricating an anchor) is worse.

File: lib/prompt.ml [hunk 5]
```diff
@@ -493,7 +493,9 @@ Do not repeat information already in the description or specs above — add only
 Keep it concise — a few bullet points usually suffices. If you have nothing material to add (the patch is straightforward), write a single line acknowledging that.|}
         vars)
 
-let render_review_prompt ~(project_name : string) ?pr_number
+let short_sha s = if String.length s <= 7 then s else String.sub s ~pos:0 ~len:7
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:497]: Added a `(* Private helper — … *)` comment. Keeping it out of the `.mli` for now since no caller needs it; the comment flags the intent so it's not accidentally wired up elsewhere.

File: lib/prompt.ml [hunk 6]
```diff
@@ -520,27 +533,75 @@ let render_review_prompt ~(project_name : string) ?pr_number
...
+               If a comment is marked `[outdated]` or refers to code that no \
+               longer exists at HEAD, reply acknowledging it's addressed in a \
+               later commit and skip — do not re-do the change.\n"
+              anchored (short_sha head)
+        | _ -> ""
+      in
+      let reviewed_at_sha_var =
+        match reviewed_at_shas with
+        | [ sha ] -> short_sha sha
+        | _ :: _ as many -> List.map many ~f:short_sha |> String.concat ~sep:","
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml]: Fixed — `reviewed_at_sha_var` now joins with `", "` to match the `sha_anchor` preamble formatting. Custom prompt templates using `{{reviewed_at_sha}}` for a multi-SHA batch will now get a properly spaced list.

File: lib/prompt.ml [hunk 7]
```diff
@@ -558,7 +619,7 @@ let render_review_prompt ~(project_name : string) ?pr_number
              { isResolved } } }'`\n\n\
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:634]: Fixed — dropped the trailing `\n` from the `sha_anchor` Printf template so the format-literal `\n\n` between `%s%s` and "The following" produces exactly one blank line. Updated the expect-test accordingly (one blank line between preamble and body instead of two).

File: lib/prompt.ml [hunk 8]
```diff
@@ -877,3 +942,105 @@ let%test "review prompt formats comments" =
   && String.is_substring result ~substring:"General feedback."
   && String.is_substring result ~substring:"resolveReviewThread"
   && String.is_substring result ~substring:"gh api repos/"
+  (* Back-compat: no SHAs → no preamble, no [at=…], no [outdated]. *)
+  && (not (String.is_substring result ~substring:"Review anchored at"))
+  && (not (String.is_substring result ~substring:"[at="))
+  && not (String.is_substring result ~substring:"[outdated]")
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:967]: Added an inline comment explaining why this plain-substring check is safe in this test context (no preamble is emitted, so the literal `[outdated]` in the preamble text never lands in the output) and pointing future editors to the bullet-line-scoped pattern used by the file-level-comment test below.

File: lib/prompt.ml [hunk 9]
```diff
@@ -520,27 +534,78 @@ let render_review_prompt ~(project_name : string) ?pr_number
...
+          ("reviewed_at_sha", reviewed_at_sha_var);
+          ("current_head_sha", current_head_sha_var);
+          ("sha_anchor", sha_anchor);
         ]
       in
       render_with_override ~project_name ~name:"review" ~vars
         ~default:(fun () ->
           Printf.sprintf
-            "# Review Comments%s\n\n\
+            "# Review Comments%s%s\n\n\
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:619]: Good catch — the double blank line is real. Fixed it by stripping the trailing `\n` from `pr_ctx` instead of touching `sha_anchor`. The reason: your suggested fix (sha_anchor: `\n\n` → `\n`) would collapse the blank line between `# Review Comments` and the preamble in the no-`pr_number` case (pr_ctx=""), since the format-literal only supplies the leading separator once. Stripping pr_ctx's trailing `\n` instead means both blocks contribute their own `\n\n` leading separator and no trailing newline, so all four subsets {pr_ctx, sha_anchor, both, neither} render with exactly one blank line between adjacent non-empty blocks. Added an expect-test covering the both-present case to lock this in.

File: lib/prompt.ml [hunk 10]
```diff
@@ -520,27 +534,78 @@ let render_review_prompt ~(project_name : string) ?pr_number
               | Some tid -> Printf.sprintf " [thread_id=%s]" tid
               | None -> ""
             in
-            Printf.sprintf "- **Comment**%s%s%s: %s" location db_id thread_ref
-              c.Comment.body)
+            let anchor =
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:542]: Added the clarifying comment — spells out that `[at=…]` shows `original_commit_sha` (the reviewer's diff anchor) rather than `commit_sha` (where GitHub re-anchored), because the agent needs to know which version of the code the reviewer was looking at.

File: lib/prompt.ml [hunk 11]
```diff
@@ -877,3 +946,108 @@ let%test "review prompt formats comments" =
...
+    |}]
+
+let%test "review prompt marks outdated comments" =
+  let comments : Comment.t list =
+    [
+      Comment.
+        {
+          id = Comment_id.of_int 1;
+          thread_id = Some "PRRT_outdated";
+          body = "This line moved.";
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/prompt.ml:1018]: Replaced the substring-based %test with a %expect_test that pins the full rendered output, so any change to the bullet ordering (`[at=…]` vs `[outdated]` position) surfaces as a diff. Also added a second %expect_test specifically covering the pr_number + current_head_sha combination to lock in the blank-line fix from the sibling comment.

File: lib/github.ml [hunk 12]
```diff
@@ -135,6 +136,8 @@ let graphql_query =
               body
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/github.ml:138]: Added a # comment inside the GraphQL query noting the opener-only constraint, so a future PR that tries to surface per-reply SHAs hits the documentation before the surprise.
